Add libholoscan-dev for C++ SDK

billysuh7 opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [X] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [X] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

**Note**: please do not approve this PR as the version will be bumped later when the next version is GA - this is an early attempt at getting this working and reviewed.

This PR enables a new subpackage (`libholoscan-dev`) to support C++ SDK usage case, which, until now, had been excluded from release due to internal issues.

I've moved `examples/` folder back into the `libholosca-dev` and removed the removal of revended `lib/libyaml-cpp*` because holoscan team has removed `libyaml` from the source tarball.

Also removed `README.md` from the package and instead created `post-link.sh` to refer to `README.md` for how to use holoscan utility scripts under `$CONDA_PREFIX/bin`

We won't be shipping `libyaml-cpp.a` but including it for cmake test to pass.

**Note**: please do not approve this PR as the version will be bumped later when the next version is GA - this is an early attempt at getting this working and reviewed.

<!--
Please add any other relevant info below:
-->

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14919911709](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14919911709). Examine the logs at this URL for more detail.</sub>

billysuh7 commented:
One thing I cannot figure out is why there are many instances of this type of error:

```
2025-05-09T02:25:19.4735661Z ERROR:conda.stderr:
2025-05-09T02:25:19.4740673Z SafetyError: The package for libholoscan located at /home/conda/feedstock_root/build_artifacts/pkg_cache/libholoscan-3.2.0.2-h6877135_0
2025-05-09T02:25:19.4741607Z appears to be corrupted. The path 'lib/libgxf_app.so'
2025-05-09T02:25:19.4741854Z has an incorrect size.
2025-05-09T02:25:19.4742060Z   reported size: 11399505 bytes
2025-05-09T02:25:19.4742591Z   actual size: 10199184 bytes
```

These do not fail the CI but it bothers me. Could someone see what can be done about this ? Some data points:

- Such libs all come from `libholoscan` subpackage, but these errors happen during the build for `libholoscan-dev` subpackage, which depends on `libholoscan`. It is as if `libholoscan-dev` build is rechecking these files where it should not..?
- When I just build `libholoscan` (without building `libholoscan-dev`) then no such errors are seen
- The `actual size` is the original size from source tarball; The `reported size` is the size of the modified after binary relocation. This seems counter-intuitive. After installing the resulting package, I confirmed the actual size matches that of `reported size`. So the conda build is somehow getting the `actual size` from the source tarball.

carterbox commented:
I believe this is a bug unique to multi-output recipes caused by the way that the PREFIX is cached for multi-output builds. When building libholoscan-dev, libgxf_app.so is first installed to the PREFIX as a dependency from the conda package, then it is overwritten with libgxf_app.so from the top-level build cache. Finally, when conda tried to validate that the PREFIX files are not corrupt, it sees that libgxf_app.so is the unpatched version from the top-level build instead of the patched one from the conda package.

billysuh7 commented:
> I believe this is a bug unique to multi-output recipes caused by the way that the PREFIX is cached for multi-output builds. When building libholoscan-dev, libgxf_app.so is first installed to the PREFIX as a dependency from the conda package, then it is overwritten with libgxf_app.so from the top-level build cache. Finally, when conda tried to validate that the PREFIX files are not corrupt, it sees that libgxf_app.so is the unpatched version from the top-level build instead of the patched one from the conda package.

Thanks @carterbox , that is good to know. After installing `libholoscan-dev` / `libholoscan`, I confirmed that all the libs had the size matching the `reported size` (not `actual size`), so based on this I agree that it must be a bug and there is nothing wrong with the libs.

carterbox commented on recipe/meta.yaml:
```diff
@@ -49,6 +49,7 @@ outputs:
         - libnpp-dev  # libnpp is tied to specific cuda-version at x.x, so relax it, as we support cuda-version >= 12.6, < 13
+      post-link: post-link
     files:
       - NOTICE/NOTICE
```
This can go in the license section instead of being installed to PREFIX? Notice/notice is a very generic filename that could be clobbered by other packages.

carterbox reviewed (changes requested):
- [ ] It seems that `yaml-cpp` is not being used by `holoscan` at runtime, but it is listed in the host requirements. Is it no longer used by the python bindings or is it being statically linked?
- [ ] The `libholoscan` package has a bunch of unused runtime dependencies on CTK package. Are these packages dlopened?

billysuh7 commented on recipe/meta.yaml:
```diff
@@ -49,6 +49,7 @@ outputs:
       ignore_run_exports_from:
         - {{ compiler('cuda') }}
         - libnpp-dev  # libnpp is tied to specific cuda-version at x.x, so relax it, as we support cuda-version >= 12.6, < 13
```
It works! Done.

billysuh7 commented:
> * [ ]  It seems that `yaml-cpp` is not being used by `holoscan` at runtime, but it is listed in the host requirements. Is it no longer used by the python bindings or is it being statically linked?

I'll get back to you on this, why it says it is no longer used when we meant to replace their revended `yaml-cpp` libs.

> * [ ]  The `libholoscan` package has a bunch of unused runtime dependencies on CTK package. Are these packages dlopened?

Thanks for catching! I can drop `rdma-core`, `libzlib`, `llvm-openmp` as no longer needed. But for the remainders, l'll get back to you - holoscan team originally wanted to include as much of the same dependencies are specified for its debian counterpart which had those dependencies, either for end user convenience or maybe some other reasons like dlopen.

billysuh7 commented:
> > * [X]   It seems that `yaml-cpp` is not being used by `holoscan` at runtime, but it is listed in the host requirements. Is it no longer used by the python bindings or is it being statically linked?
> 
> I'll get back to you on this, why it says it is no longer used when we meant to replace their revended `yaml-cpp` libs.
> 

`yaml-cpp` is static linked now but it is still part of `holoscan` public interface so I moved `yaml-cpp` from `holoscan` package to `libholoscan-dev` package, which still needs yaml headers

> > * [X]   The `libholoscan` package has a bunch of unused runtime dependencies on CTK package. Are these packages dlopened?
> 
> Thanks for catching! I can drop `rdma-core`, `libzlib`, `llvm-openmp` as no longer needed. But for the remainders, l'll get back to you - holoscan team originally wanted to include as much of the same dependencies are specified for its debian counterpart which had those dependencies, either for end user convenience or maybe some other reasons like dlopen.

Added `cupy` and eliminated many dependencies covered by `cupy` per suggestion from holoscan team.
Reference: https://github.com/nvidia-holoscan/holoscan-sdk/blob/main/cmake/modules/HoloscanCPack.cmake#L70-L101

Also checked w/ holoscan team to take out `libnuma1` package for `linux-aarch64` which does not seem to need it anymore.

With the last commit I no longer see any unused dependency warnings.

carterbox commented on recipe/meta.yaml:
```diff
@@ -62,44 +63,28 @@ outputs:
-        - cuda-cudart-dev
         - cuda-version {{ cuda_version }}
+        - cuda-nvrtc-dev
+        - cupy
```
> Added cupy and eliminated many dependencies covered by cupy per suggestion from holoscan team.

cupy cannot be a dependency of the C-API package. That doesn't make any sense because cupy is a Python library.

This recipe should list only its own direct dependencies. It shouldn't use cupy as dependency proxy.

billysuh7 replied:
Hi @carterbox that totally makes sense; I've moved `cupy` under `holsocan` usbpackage (unrelated to this, added `cuda-nvcc`).

billysuh7 commented:
@conda-forge-admin, please restart ci
